### PR TITLE
Added rgb to bgra conversion

### PR DESCRIPTION
**Pull request by Vendrik**
Added rgb to bgra conversion including rust and c bindings tests

### CODE REVIEW

File: src/convert_image/x86.rs [hunk 1]
```diff
@@ -16,6 +16,38 @@
 
 use crate::convert_image::common::*;
 
+#[cfg(target_arch = "x86")]
+use core::arch::x86::_bswap;
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: Is there a particular reason to import from core crate?
Is it possible to import from:
`std::arch`

like for the other conversion functions?

File: src/convert_image/x86.rs [hunk 2]
```diff
@@ -16,6 +16,38 @@
...
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::_bswap;
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::_bswap64;
+
+#[cfg(target_arch = "x86_64")]
+#[inline(always)]
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: I think rust is able to inline automatically this function in release.
In this case it is preferable to leave the function without the inline.

File: src/convert_image/x86.rs [hunk 3]
```diff
@@ -621,3 +653,120 @@ pub fn nv12_bt709_bgra_lrgb(
...
+    const SHIFT_40: u8 = 40;
+    const SHIFT_16: u8 = 16;
+    const SHIFT_8: u8 = 8;
+
+    let w = width as usize;
+    let output_stride_diff = if dst_strides[0] == 0 {0} else {dst_strides[0] - (OUTPUT_BPP * w)};
+    let input_stride_diff = if src_strides[0] == 0 {0} else { src_strides[0] - (INPUT_BPP * w)};
+
+    /*
+     * For single swap iteration, since the swap is done on 32 bits while the input is only
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: Rust comments preferable with `//`

File: src/convert_image/x86.rs [hunk 4]
```diff
@@ -16,6 +16,38 @@
...
+
+#[cfg(target_arch = "x86_64")]
+#[inline(always)]
+unsafe fn byte_swap64 (x: i64) -> i64 {
+    _bswap64(x)
+}
+
+#[cfg(target_arch = "x86")]
+#[inline(always)]
+unsafe fn byte_swap32 (x: i32) -> i32 {
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: Same here

File: src/convert_image/x86.rs [hunk 5]
```diff
@@ -16,6 +16,38 @@
...
+    _bswap(x)
+}
+
+#[cfg(target_arch = "x86")]
+#[inline(always)]
+unsafe fn byte_swap64 (x: i64) -> i64 {
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: Same here

File: src/convert_image/x86.rs (hunk 1 above)
Comment (Vendrik) [reply src/convert_image/x86.rs]: By looking at rust docs it seems that _bswap functions are available only in core::arch, there is a bswap function in std::intrinsics but is experimental

File: benches/benches.rs [hunk 6]
```diff
@@ -185,6 +180,56 @@ fn nv12_bgra(
     Ok(elapsed)
 }
 
+fn rgb_bgra(mut input_file: &mut Cursor<&[u8]>, output_path: &str) -> BenchmarkResult<Duration> {
+    let (mut width, height, input_buffer) = { pnm_data(&mut input_file)? };
+    width /= 3;
+    //println!("w={}", width);
```
Comment (fabiosky) [inline benches/benches.rs]: Remove this, I think it was accidentally left

File: benches/geninput.py [hunk 7]
```diff
@@ -1,12 +1,15 @@
...
+from itertools      import product as cartesian_product
+from random         import Random
+from os.path        import join, exists, dirname, realpath
+from os             import environ
 import sys
 
+sys.path.append(dirname(realpath(__file__)))
+
+from fractions import Fraction as frac
+
```
Comment (fabiosky) [inline benches/geninput.py]: Those lines should not be changed:
1. Python pep8 do not want additional spaces after 'from x' and 'import'
2. Fractions module is installed using pip

File: benches/geninput.py [hunk 8]
```diff
@@ -102,9 +123,7 @@ def chunks(l, n):
         CbCr = array('B', [C_HALF] * (ATLAS_WIDTH * height // 2))
         i = 0
         j = 0
-
-        # half because y has to be repeated
-        for samples in chunks(sample_list, ATLAS_WIDTH // 2):
+        for samples in chunks(sample_list, ATLAS_WIDTH // 2): # half because y has to be repeated
```
Comment (fabiosky) [inline benches/geninput.py]: Unnecessary change

File: c_tests/unit.c [hunk 9]
```diff
@@ -1203,6 +1203,90 @@ unit_convert_image_over_4gb_limit(void)
...
+    DcpStatus status = dcp_status();
+    TEST_BEGIN_GROUP(__FUNCTION__);
+    init();
+    for (width = 0; width < MAX_WIDTH; width++) {
+        for (height = 0; height < MAX_HEIGHT; height++) {
+            for (src_stride_diff = 0; src_stride_diff < MAX_FILL_BYTES; src_stride_diff++) {
+                for (dst_stride_diff = 0; dst_stride_diff < MAX_FILL_BYTES; dst_stride_diff++) {
+                    uint32_t h;
+                    uint32_t w;
+                    TEST_BEGIN("Width=%d, Height=%d, Stirdes: src=%d, dst=%d", width, height, src_stride_diff, dst_stride_diff);
```
Comment (fabiosky) [inline c_tests/unit.c]: Typo: stirdes

File: c_tests/unit.c [hunk 10]
```diff
@@ -1203,6 +1203,90 @@ unit_convert_image_over_4gb_limit(void)
...
+                    dst_stride_bytes = (dst_bpp * width) + dst_stride_diff;
+                    src_stride_param = src_stride_diff == 0 ? auto_stride : src_stride_bytes;
+                    dst_stride_param = dst_stride_diff == 0 ? auto_stride : dst_stride_bytes;
+
+                    src_buffers[0] = stack_alloc(src_stride_bytes * height);
+                    dst_buffers[0] = stack_alloc(dst_stride_bytes * height);
+
+                    for (h = 0; h < height; h++) {
+                        for (w = 0; w < width; w++) {
+                            src_buffers[0][(h * src_stride_bytes) + (w * src_bpp) + 0] = rand() % 256;
```
Comment (fabiosky) [inline c_tests/unit.c]: Use 'rand & 0xFF'

File: src/convert_image/x86.rs [hunk 11]
```diff
@@ -621,3 +633,140 @@ pub fn nv12_bt709_bgra_lrgb(
...
+    width: u32,
+    height: u32,
+    _last_src_plane: u32,
+    src_strides: &[usize],
+    src_buffers: &[&[u8]],
+    _last_dst_plane: u32,
+    dst_strides: &[usize],
+    dst_buffers: &mut [&mut [u8]],
+) -> bool {
+    if width == 0 || height == 0 {
```
Comment (fabiosky) [inline src/convert_image/x86.rs]: We should detect when there is not enough data in the buffers, like the other algorithms.

**Comment by Vendrik** [comment]
#7
